### ISSUE TITLE
Reduce LSA object descriptor overhead

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2032]
```
        static constexpr uint8_t live_flag = 0x01;
        static constexpr uint8_t eos_flag = 0x02;
        static constexpr size_t max_alignment = (0xff >> 2) + 1;

    class object_descriptor {
    private:
        obj_flags _flags;
        uint8_t _alignment;
        segment::size_type _size;
        allocation_strategy::migrate_fn _migrator;
```

If we

 - force alignment to always be 8
 - add `migrate_fn_type::size(void* object)`
 - add a `dead_migrate_fn_type` subclass (instead of live_flag)

then we can reduce `sizeof(object_descriptor)` from 14 bytes to 8 bytes

**Comment by tgrabiec** [comment]
Rather than forcing the alignment, shouldn't we pack the objects (require alignment of 1) to reduce overhead? I've heard that nowadays misaligned accesses are no longer detrimental to performance on x86.

**Comment by avikivity** [comment]
We can force alignment of 1.  Will require adding `static_assert(alignof(T) == 1)` to make sure we declare everything as packed (not that it makes a difference).

**Comment by tgrabiec** [comment]
What do you mean by "not that it makes a difference"?

**Comment by avikivity** [comment]
It is illegal to construct an object with alignof greater than actual alignment. On x86 and aarch64, however, the same code is emitted when when constructing a regular object and a [[gnu::packed]] object, so even if you make a mistake, no one notices (except perhaps a sanitizer).

On other architectures (like armv7) you have different code emitted for the two cases.

**Comment by duarten** [comment]
Fixed by 6d9e18fd612d33c0df317d40dd10245995f5f14b, no?

**Comment by avikivity** [comment]
Yes.
